### ISSUE TITLE
Ensure that simple field-level metadata updates are propagated to downstream DCP components.

### ISSUE THREAD

**Issue by tburdett** [HumanCellAtlas/dcp#222]
# Value proposition from the Q2 roadmap

Data contributors can provide updates to their data and metadata and see those updates reflected, end to end, across the whole DCP

# Assumptions and Background

Data submitted to the DCP will need to be updated due to publications, improved ontology annotation and other new information or to fix errors in the data or metadata discovered after submission. These updates need to be captured by and reflected in the metadata stored by the DCP and distributed to the community.

A full set of features to enable the Addition, Update, Deletion and Redaction (AUDR) of and metadata in the DCP has a high engineering cost. What is presented here represents the first iteration which enables addition and update of simple metadata fields which do not have consequences for downstream analysis of the submitted data.

# Proposed Scope       

## Supported AUDR v0.1 operations:

1. **Addition of new scientific content metadata fields** to the existing data such as a new publication, a new contributor, or adding a value to an existing blank metadata field that doesn’t change how an analysis pipeline is run
2. **Changing the value of an existing scientific content metadata** field such as improving the ontology term mapping that doesn’t change how an analysis pipeline is run
3. **Changing a field in response to a schema update** such as a field name change or new required field in the schema

NB operation 3 is a stretch goal and may be moved to out of scope pending task estimates from @justincc and @norman and any other effected components.

## Individual component scoping

1. Schema migrations which are required to enable updates will be driven by wranglers using scripts that run outside of the ingest system
2. The pipeline execution service should be able to ignore any notifications about bundles which have already been analysed. Ingest will handle analysis bundle updates alongside input bundle updates.

## Out of scope AUDR v0.1 operations:

1. **Updates to the experiment graph**, ie new protocols, new biomaterials, identification of incorrect relationships between existing biomaterials, processes and files. 
2. **Updates which require additional or updated experimental data files **
3. **Updates which require an analysis pipeline to be run again**, either a new pipeline or the same pipeline with different parameters
4. An extended list can be found in the[ AUDR Pre-RFC](https://docs.google.com/document/d/1XeAw05RAJe20Q1XDF7eLWLGF3gCme4-8l-ZfhtDWOLg/edit#heading=h.pu1uyfj25v2g)

Out of scope operations will fall back on the old reingest/add old bundles to the exclusion list process which currently exists.

# Q2 Deliverables

1. Documentation for the wranglers which allows them to determine if an update will trigger a new pipeline or the same pipeline with different parameters
2. Ingest UI which supports wranglers retrieving the complete submission metadata spreadsheet for the data to be updated with added UUIDs from ingest. This spreadsheet will reflect the original schema of the submission.
3. Python script which allows wranglers to migrate old spreadsheets to new schemas
4. An ingest UI view which allows wranglers to see what documents have changed (stretch goal is the specifics of how they have changed)
5. An ingest exporter feature which can update entity JSON files and bundle JSON files in the DSS to reflect all updates/additions. This will include both input and analysis bundles.
6. Pipeline execution service feature which checks if the pipeline connected to the notification event has already run for the given bundle. If the answer is yes the pipeline execution service will stop, assuming update has been handled by ingest. If the answer is no the pipeline execution service will continue as if this was a brand new bundle. 
7. The data browser once notified of an updated bundle will recreate the index using the most recent version of each JSON and data file.
8. If the Matrix service relies on pre-calculated metadata any calculations will need to be rerun.
9. The Query service will need to reload any updated documents into their data structure.

# Future work

This represents the first step on the path to a completed AUDR feature set. Future work for subsequent quarters will include:

*   Enabling the update of experimental data files in the DSS
*   Improving interfaces to delete and retract data in the DSS
*   Wranglers being able to make simple metadata changes directly in the UI without download and reload of a spreadsheet
*   Automated schema migrations as part of release processes
*   Updates and Additions which alter the experimental graph being supported
*   Creating logic in the DCP which supports assessment for automated reanalysis triggering and cleanup of old erroneous pipeline results.

# Documents

[Use cases](https://docs.google.com/document/d/1rI8PCASomdAHznyWQceRJGv4-wtg-P5Rm-rI_uJLjtE/edit#)
[Pre RFC](https://docs.google.com/document/d/1XeAw05RAJe20Q1XDF7eLWLGF3gCme4-8l-ZfhtDWOLg/edit?ts=5beda8b0)
[Ingest High-Level Design for Updates](https://docs.google.com/document/d/1YkIqUgZStWC7mnQWHvefONWbUhAcHPDtV6rXExrWhSw/edit#heading=h.i1f06s89qmeh)

**Comment by morrisonnorman** [comment]
See the AUDR use cases document:

https://docs.google.com/document/d/1rI8PCASomdAHznyWQceRJGv4-wtg-P5Rm-rI_uJLjtE/edit#heading=h.bbk1hab0f11l

**Comment by lauraclarke** [comment]
This comment is to confirm that all responsible parties are happy with the scope and are able to start defining and estimating the implementation plan. **We should have implementation/spike plans in place by April 23rd** so they can begin to be scheduled in the **sprint starting April 30th** (assuming EBI sprint and internal Demo dates, please appropriately adjust if this is about to change).

A tick in this box means you have enough information to define an implementation plan/schedule required spikes etc. Please comment here tagging Laura/Norman/Justin/Simon if you need more information

- [ ] Ingest Support @claymfischer @morrisonnorman @malloryfreeberg 
- [ ] Upload @parthshahva @justincc 
- [ ] Data Store @briandoconnor @kozbo @xbrianh 
- [ ] Pipeline Execution @TimothyTickle @samanehsan @rexwangcc 
- [ ] Data Processing Piplines @TimothyTickle @kbergin @samanehsan @rexwangcc 
- [ ] Query Service @tjchen @mweiden @kislyuk 
- [ ] Data Browser @briandoconnor @theathorn @hannes-ucsc 
- [ ] Matrix Servive @ambrosejcarr @brianraymor @mckinsel

**Comment by justincc** [comment]
I am still working on rewriting the updates pre-RFC (then RFC) for this Q2 scope. Hope to have this done next week but if you need info sooner please ping me.

**Comment by samanehsan** [comment]
I have a few questions regarding this point under individual component scoping: 
>The pipeline execution service should be able to ignore any notifications about bundles which have already been analyzed. Ingest will handle analysis bundle updates alongside input bundle updates.

1. Does this mean we will continue copying the primary metadata into the analysis results? There was some discussion at the DCP meeting in February about removing this copy-forward behavior, since it would eliminate the need to update primary metadata in otherwise unchanged analysis bundles. I believe @justincc was looking into that as part of the AUDR pre-RFC.
2. I see this is scoped out, but what is the expectation when this is expanded to include changes that require re-analysis? If ingest handles analysis bundle updates alongside input bundle updates, it will need to know whether or not that change will result in a re-analysis. Specifically (1) which pipeline, if any, would be triggered and (2) the pipeline inputs. This could easily get out of sync as the data processing service makes changes to the analysis pipelines, subscriptions, etc. Also if the pipeline version changes, updates to a primary bundle will trigger an analysis even if the analysis inputs are unchanged.

**Comment by lauraclarke** [comment]
@samanehsan . These are great questions. This is my view on your questions @justincc @simonjupp please pipe up if I have misrepresented anything

*Does this mean we will continue copying the primary metadata into the analysis results? There was some discussion at the DCP meeting in February about removing this copy-forward behavior, since it would eliminate the need to update primary metadata in otherwise unchanged analysis bundles. I believe @justincc was looking into that as part of the AUDR pre-RFC.*

Right now there is no timeline for stopping the copy-forward of input bundle metadata into analysis bundles so this scope includes it. If it turns out through work on this epic or anything else that stopping the copy forward is desirable we would need to rescope to ingest doesn't need to update the analysis bundles when it updates the input bundles. I don't believe it would change what was needed on the pipelines side

*I see this is scoped out, but what is the expectation when this is expanded to include changes that require re-analysis? If ingest handles analysis bundle updates alongside input bundle updates, it will need to know whether or not that change will result in a re-analysis. Specifically (1) which pipeline, if any, would be triggered and (2) the pipeline inputs. This could easily get out of sync as the data processing service makes changes to the analysis pipelines, subscriptions, etc. Also if the pipeline version changes, updates to a primary bundle will trigger an analysis even if the analysis inputs are unchanged.*

My view is that ingest holding logic about if an update should trigger a reanalysis is probably a bad idea. The scope for this quarter is to minimise work on the pipelines and ingest sides and also to enable simple updates to move forward. When we reconsider scope in future quarters (this might be as soon as quarter 3 2019 depending on everyone's capacity and other priorities) then looking at building a library to sensibly diff input bundles so logic could be built to understand if a reanalysis needs to be triggered is a very sensible plan. 

One issue that I know we will need to think through very carefully is what happens if an input bundle has already been analysed once using one pipeline (say human) but then gets updated so it needs to be analysed by a different pipeline (say mouse). Can we automate or at least automatically flag for review the old analysis results for deletion

**Comment by justincc** [comment]
@samanehsan The current scope needs you to detect when a primary bundle has been analyzed and *not* reanalyze it. Also in this scope, you don't need to signal non-analysis - ingest will simply copy-forward secondary bundles itself. 

The tricky questions of how to signal new/update and tell which files have been updated that we have been discussing are now pushed into the future.

**Comment by lauraclarke** [comment]
It would be great to get ticks in the remaining boxes or to understanding what people's questions are in order to tick the boxes and make plans

@claymfischer @morrisonnorman @malloryfreeberg @parthshahva @justincc @kozbo @xbrianh @TimothyTickle @samanehsan @rexwangcc @tjchen @mweiden @kislyuk @briandoconnor @theathorn @hannes-ucsc @ambrosejcarr @brianraymor @mckinsel

**Comment by brianraymor** [comment]
Has a decision be made about whether _3. Changing a field in response to a schema update such as a field name change or new required field in the schema_ is in scope? 

My expectation from the Technical Architecture call is that Justin will [collaborate on a RFC](HumanCellAtlas/dcp#366) that will unblock this issue and allow the shepherd to create child issues that support this epic in other repo(s). We're not using _checkboxes_ in epics.

**Comment by brianraymor** [comment]
I also do not understand the [comment](https://github.com/HumanCellAtlas/dcp/issues/222#issuecomment-482028539)  _The current scope needs you to detect when a primary bundle has been analyzed and not reanalyze it, even if primary data has changed that would result in a different analysis_

because the original scope says _What is presented here represents the first iteration which enables addition and update of simple metadata fields which **do not have consequences for downstream analysis** of the submitted data._

**Comment by mweiden** [comment]
@brianraymor I agree that the RFC would be helpful, ideally focused to the requirements and design for Q2.

@lauraclarke re. bringing better definition to this work: we've left many comments/questions on @justincc's RFC draft and the tickets he's linked to from that doc. Most have not been resolved or synthesized into the plan for Q2 yet.

**Comment by justincc** [comment]
> Has a decision be made about whether _3. Changing a field in response to a schema update such as a field name change or new required field in the schema_ is in scope?
> 
> My expectation from the Technical Architecture call is that Justin will [collaborate on a RFC](HumanCellAtlas/dcp#366) that will unblock this issue and allow the shepherd to create child issues that support this epic in other repo(s). We're not using _checkboxes_ in epics.

3. now relies on a script external to ingest to convert the retained spreadsheets to the new schema. @simonjupp we agreed that you would be doing this? If so, there should be little for ingest itself to do.

**Comment by justincc** [comment]
> @brianraymor I agree that the RFC would be helpful, ideally focused to the requirements and design for Q2.
> 
> @lauraclarke re. bringing better definition to this work: we've left many comments/questions on @justincc's RFC draft and the tickets he's linked to from that doc. Most have not been resolved or synthesized into the plan for Q2 yet.

The pre-RFC now addresses the significantly reduced scope that doesn't require many of those comments/questions and tickets to be resolved. Once there is a decision on the interface between ingest and the datastore for updates then I will proceed with the actual RFC.

**Comment by brianraymor** [comment]
Regarding 3

> If so, there should be little for ingest itself to do

But there is potential impact on downstream consumers that needs to be addressed. For example, a **field name change** would have similar impact as the April [metadata style changes](https://github.com/HumanCellAtlas/metadata-schema/issues/691) and re-ingestion plan.

**Comment by mweiden** [comment]
@justincc thanks for the new Q2 section. It's much clearer.

Separate topic from @brianraymor's thread above:

> I also do not understand the comment above: The current scope needs you to detect when a primary bundle has been analyzed and not reanalyze it, even if primary data has changed that would result in a different analysis
>
> because the original scope says What is presented here represents the first iteration which enables addition and update of simple metadata fields which do not have consequences for downstream analysis of the submitted data.

If I'm understanding your proposal correctly, are you suggesting that we don't worry about these SAB updates for now?

De-scoping AUDR in Q2 just to updates that do not affect analysis outputs in Q2 might be OK. If we want to offer a good user experience, we should put guardrails in place that make sure bundles are consistent with each other. If we don't, primary bundles will likely get into a weird state in relation to secondary analysis bundles and we won't have a good way of validating that the data is consistent.

**Comment by justincc** [comment]
> Regarding 3
> 
> > If so, there should be little for ingest itself to do
> 
> But there is potential impact on downstream consumers that needs to be addressed. For example, a **field name change** would have similar impact as the April [metadata style changes](https://github.com/HumanCellAtlas/metadata-schema/issues/691) and re-ingestion plan.

It should be the same impact as the existing plan, unless wranglers choose to make additional schema updates.

**Comment by justincc** [comment]
> @justincc thanks for the new Q2 section. It's much clearer.
> 
> Separate topic from @brianraymor's thread above:
> 
> > I also do not understand the comment above: The current scope needs you to detect when a primary bundle has been analyzed and not reanalyze it, even if primary data has changed that would result in a different analysis
> > because the original scope says What is presented here represents the first iteration which enables addition and update of simple metadata fields which do not have consequences for downstream analysis of the submitted data.
> 
> If I'm understanding your proposal correctly, you're suggesting that we don't worry about these SAB updates for now?
> 
> De-scoping AUDR in Q2 just to updates that do not affect analysis outputs in Q2 might be OK, but if we want to offer a good user experience, we should put guardrails in place that make sure primary and secondary analysis bundles are consistent with each other until updates propagate through secondary analysis. If we don't, primary bundles will likely get into a weird state in relation to secondary analysis bundles and we won't have a good way of validating that the data is consistent.

Yes, the proposal is that performing secondary analysis updates will be out of scope for Q2.

The primary and secondary bundles should remain consistent because we are going to update both directly. In other words, for Q2 we will not be relying on any notification from analysis that they have not chosen to update the analysis.

**Comment by lauraclarke** [comment]
> Has a decision be made about whether _3. Changing a field in response to a schema update such as a field name change or new required field in the schema_ is in scope?
> 
> My expectation from the Technical Architecture call is that Justin will [collaborate on a RFC](HumanCellAtlas/dcp#366) that will unblock this issue and allow the shepherd to create child issues that support this epic in other repo(s). We're not using _checkboxes_ in epics.

@brianraymor is there a prefered solution for transparent confirmation that the responsible components are happy with the prefered scope. I was copying the checkbox solution from the wranglers as it seems like a low cost way to do transparent asynchronous positive confirmation without having to create a lot of extra tickets

**Comment by lauraclarke** [comment]
> > @justincc thanks for the new Q2 section. It's much clearer.
> > Separate topic from @brianraymor's thread above:
> > > I also do not understand the comment above: The current scope needs you to detect when a primary bundle has been analyzed and not reanalyze it, even if primary data has changed that would result in a different analysis
> > > because the original scope says What is presented here represents the first iteration which enables addition and update of simple metadata fields which do not have consequences for downstream analysis of the submitted data.
> > 
> > 
> > If I'm understanding your proposal correctly, you're suggesting that we don't worry about these SAB updates for now?
> > De-scoping AUDR in Q2 just to updates that do not affect analysis outputs in Q2 might be OK, but if we want to offer a good user experience, we should put guardrails in place that make sure primary and secondary analysis bundles are consistent with each other until updates propagate through secondary analysis. If we don't, primary bundles will likely get into a weird state in relation to secondary analysis bundles and we won't have a good way of validating that the data is consistent.
> 
> Yes, the proposal is that performing secondary analysis updates will be out of scope for Q2.
> 
> The primary and secondary bundles should remain consistent because we are going to update both directly. For Q2 we will not be relying on any notification from analysis that they have not chosen to update the analysis.

To add to @justincc comments

The idea behind asking the pipelines team to be able to check if an analysis has already been run and ignore it has two aspects. 

The first is foundational, eventually, the pipeline execution service will be able to catch and made decisions about whether to reanalyse updated data or not and that logic should lay with the pipelines team rather than ingest. This is a first step in that process and one that @samanehsan has previously confirmed that this is a possible extension to their existing system

The second is pragmatic, my understanding of the eventing service is that updated bundled will trigger notification events and catching and handling them in pipelines was always part of the original design. Changing that approach now seems like it creates more work for more people.

We are making an assumption that ingest can ensure that with the current copy forward scenario all input metadata in the analysis bundles can be updated by ingest and that the pipelines team can provide sufficient documentation to the wranglers team that in the case where a reanalysis would actually be triggered by an update we fall back to the old process. We will have to try the process to make sure it works but as we are trying to be quite ruthless with scope so the first improvement to AUDR happens that felt like a reasonable position to take right now.

@mweiden very happy to consider what are the best guardrails to ensure that we don't create confusion for our data consumers. Do you have proposals on how we might do that?

**Comment by lauraclarke** [comment]
> > Regarding 3
> > > If so, there should be little for ingest itself to do
> > 
> > 
> > But there is potential impact on downstream consumers that needs to be addressed. For example, a **field name change** would have similar impact as the April [metadata style changes](https://github.com/HumanCellAtlas/metadata-schema/issues/691) and re-ingestion plan.
> 
> It should be the same impact as the existing plan, unless wranglers choose to make additional schema updates.

**Comment by lauraclarke** [comment]
apologies, clicked the wrong button

**Comment by lauraclarke** [comment]
> > Has a decision be made about whether _3. Changing a field in response to a schema update such as a field name change or new required field in the schema_ is in scope?
> > My expectation from the Technical Architecture call is that Justin will [collaborate on a RFC](HumanCellAtlas/dcp#366) that will unblock this issue and allow the shepherd to create child issues that support this epic in other repo(s). We're not using _checkboxes_ in epics.
> 
> @brianraymor is there a prefered solution for transparent confirmation that the responsible components are happy with the proposed scope. I was copying the checkbox solution from the wranglers as it seems like a low cost way to do transparent asynchronous positive confirmation without having to create a lot of extra tickets

Just to extend this comment. I am concerned that if all other components planning for this ticket wait for the RFC rather than considering and asking questions based on the scope description that we will have a long wait before implementation work starts and risk not delivering a finished product by the end of the quarter.

**Comment by sampierson** [comment]
@justincc @lauraclarke I have been asked to add my 2c here.  The bureaucrats are unhappy because you’re proposing to not follow their RFC process.

On the one hand I agree that our RFC process is unwieldy and can be a giant time suck.   Some folks are overly fond of debate, and seem to have nothing better to do than throw corner-case “what-if”s at you all day that will probably never happen in the real world.

On the other hand, an RFC, if done right, would be just a differently shaped vehicle you can used to drive the outstanding questions in this design to ground, and create consensus.

I would suggest the following:

1. I suggest you write an RFC.  But here hear me out…
2. I think you already have everything you need for an RFC here.   Not the AUDR RFC, but a simpler one perhaps entitled something like “Primary Metadata updates Interim Solution”.   Start with the description at the top of this ticket, copy in everything above “Historical discussion” from the “Primary Metadata updates pre-RFC” document, then copy in some of the urgent use-cases from “Project addition, update, deletion, and retraction use cases”, and call it good.  Don't spend a lot of time on it.  You should be able to knock this together in an hour.  Feel free to leave open questions in it, to be answered during the review process.
3. Publish your RFC immediately to start the clock ticking.  I am told that RFCs should take only 3 weeks, 2 weeks for community review and one week for oversight review (obviously this ignores the several months of research, design and debate required to get an RFC written in the first place).
4. To prevent the RFC time-suck, schedule two 1-hour review meetings for discussion (one a week).  Refuse to engage in lengthly comment tangents in the doc.  Tell people to come to your meetings for long discussions.
4. Start work on your implementation already.  You don’t need to have the RFC approved before you can write a line of code.  If its trajectory changes dramatically yes there will be some rework involved, but the opportunity cost is too great to wait.

**Comment by brianraymor** [comment]
Validating that _3. Changing a field in response to a schema update such as a field name change or new required field in the schema_ is now out-of-scope since I no longer see it in **Supported AUDR v0.1 operations**.

**Comment by mweiden** [comment]
@lauraclarke 

> @mweiden very happy to consider what are the best guardrails to ensure that we don't create confusion for our data consumers. Do you have proposals on how we might do that?

There are multiple ways of addressing the core concern of ensuring primary and secondary bundles remain consistent with each other. Some possibilities: making updates contingent on assertions that a pipeline is not started, rollback in the case that one is started, re-ingestion, programmatically checking pipeline inputs... Choosing which one should really be left up to the tech leads and engineers leading the effort. What _is_ important is that this concern is addressed in a reasonable/scalable way so we provide a good user experience.

**Comment by lauraclarke** [comment]
Thanks, @sampierson for the feedback. We are drafting the short document now. Given the rapidly approaching public holidays due to Easter, it may be it doesn't get into a PR before the middle of next week.

**Comment by lauraclarke** [comment]
> @lauraclarke
> 
> > @mweiden very happy to consider what are the best guardrails to ensure that we don't create confusion for our data consumers. Do you have proposals on how we might do that?
> 
> There are multiple ways of addressing the core concern of ensuring primary and secondary bundles remain consistent with each other. Some possibilities: making updates contingent on assertions that a pipeline is not started, rollback in the case that one is started, re-ingestion, programmatically checking pipeline inputs... Choosing which one should really be left up to the tech leads and engineers leading the effort. What _is_ important is that this concern is addressed in a reasonable/scalable way so we provide a good user experience.

Thanks @mweiden those are great suggestions. So far I had been thinking more about how the wranglers would know if a reanalysis would be needed before starting the update process. I think some sort of automated check to ensure that we don't have duplicate analysis results for the same input data is a great idea.

**Comment by justincc** [comment]
> @lauraclarke
> 
> > @mweiden very happy to consider what are the best guardrails to ensure that we don't create confusion for our data consumers. Do you have proposals on how we might do that?
> 
> There are multiple ways of addressing the core concern of ensuring primary and secondary bundles remain consistent with each other. Some possibilities: making updates contingent on assertions that a pipeline is not started, rollback in the case that one is started, re-ingestion, programmatically checking pipeline inputs... Choosing which one should really be left up to the tech leads and engineers leading the effort. What _is_ important is that this concern is addressed in a reasonable/scalable way so we provide a good user experience.



> > @lauraclarke
> > > @mweiden very happy to consider what are the best guardrails to ensure that we don't create confusion for our data consumers. Do you have proposals on how we might do that?
> > 
> > 
> > There are multiple ways of addressing the core concern of ensuring primary and secondary bundles remain consistent with each other. Some possibilities: making updates contingent on assertions that a pipeline is not started, rollback in the case that one is started, re-ingestion, programmatically checking pipeline inputs... Choosing which one should really be left up to the tech leads and engineers leading the effort. What _is_ important is that this concern is addressed in a reasonable/scalable way so we provide a good user experience.
> 
> Thanks @mweiden those are great suggestions. So far I had been thinking more about how the wranglers would know if a reanalysis would be needed before starting the update process. I think some sort of automated check to ensure that we don't have duplicate analysis results for the same input data is a great idea.

Consistency will be assured by the fact that ingest will be doing all updates at this point (primary and secondary). We are not relying on a trigger from green box as there is none in this design.

**Comment by brianraymor** [comment]
Assuming that the scope is limited to 1 and 2 above (and **not** 3), then the matrix service can respond to notifications of new bundle versions as illustrated in the [Cross-DCP diagram](https://docs.google.com/document/d/1XeAw05RAJe20Q1XDF7eLWLGF3gCme4-8l-ZfhtDWOLg/edit#heading=h.usajl1m430qg).

However, the matrix service does not support multiple versions of bundles (it will replace the previous version with the new version), so there would need to be agreement/alignment with the data browser (@NoopDog and @theathorn) on the same bundle versions in this case.

It would also be helpful to have a documented process for announcing changes so we can validate the success of these operations.

**Comment by malloryfreeberg** [comment]
> It would also be helpful to have a documented process for announcing changes so we can validate the success of these operations.

@brianraymor I don't know what breadth of "these operations" you are thinking, but announcing when updated primary data has successfully been exported to the DSS could be done by wranglers on the public #data-wrangling channel. We have done this in the past on-and-off, but I think wranglers would generally be ok with agreeing to do this.

**Comment by brianraymor** [comment]
@malloryfreeberg - yesterday, the matrix service was discussing how we're having difficulty tracking the [11 weeks of re-ingestion](https://humancellatlas.slack.com/archives/C9ENBPVNW/p1555428629011400) despite the existence of the re-ingestion plan. A re-ingestion calendar that was kept updated would be helpful for example.

**Comment by diekhans** [comment]
Given the current bundle-level event mechanism, there is no way
to ensure consistency beyond the bundle.  As soon as one bundle
is updated, events will fire.  There is no guarantee of time
order. Secondary bundle updates maybe seen before primary bundle
updates.  Bundles may not be consistent across a project when
analysis runs.

This lack of any type of ACID guarantee beyond the bundle
creates a lot of complexity for the downstream consumers to
determine the state of any group of bundles.

Any attempt at doing synchronization by assumptions is going to
have holes in it.

It is also very difficult to derive any data that cross bundle
boundaries.  For instance, it is not tractable to build a
project level metadata.json based on bundle event updates.
There is both theoretical and empirical evidence that this would
take weeks or longer for larger projects.

I believe we need to consider common method of doing
synchronization beyond the individual bundle rather than let
ever component make it up for themselves.

**Comment by malloryfreeberg** [comment]
@brianraymor There is a schedule listed in this [doc](https://docs.google.com/document/d/1qjsVwfLx4-Dd0NnQ0-ZB0mTexK8nZXqELmutYooWZG0/edit#). I don't think making calendar events will be useful here, as we don't have a specific date/time we are re-ingesting.

**Comment by lauraclarke** [comment]
@sampierson The RFC now exists

https://github.com/HumanCellAtlas/dcp-community/pull/77

**Comment by justincc** [comment]
Now waiting for simple updates testing to be approved in the DCP-wide integration test https://humancellatlas.slack.com/archives/C6SGU0R3J/p1564589623057300
